text/0031-integ-testing.md

```diff
@@ -58,39 +56,44 @@ The following integ test invokes puts an object in the S3 bucket and verifies th
 ```ts
 // repo.contest.ts
 import { ArtifactRepo } from './records';
-import { Test } from '@aws-cdk/contest';
+import { AwsAssertionCall, AwsAssertion, Test } from '@aws-cdk/contest';
 
 const repo = new ArtifactRepo(app, 'RequestRecord');
 
 const repoTest = Test.forConstruct(repo);
 
-repoTest.assertAws.
-  s3.putObject({ bucket: stack.artifacts, key: '...', body: '...' });
+new AwsAssertion(repoTest, 'RepoPutObject', {
+  request: AwsAssertionCall.S3.putObject({ bucket: stack.artifacts, key: '...', body: '...' })
+});
 
-repoTest.assertAws.
-  sqs.receiveMessage({ queue: stack.publishNotifs })
-     .returns({ Messages: [ ... ] });
+new AwsAssertion(repoTest, 'MessageReceived', {
+  request: AwsAssertionCall.Sqs.receiveMessage({ queue: stack.publishNotifs }),
+  returns: { Messages: [...] }
+});
 ```
 
 ### Assertions
 
 `contest` comes canned with a set of commonly used high level assertions.
 
-* `assertAws`: execute AWS service APIs and assert the response.
-* `invokeLambda`: invoke an AWS lambda function and assert the response.
-* `httpRequest`: Execute an HTTP request against an API endpoint and assert the response.
+* `AwsAssertion`: execute AWS service APIs and assert the response.
+* `InvokeLambda`: invoke an AWS Lambda Function and assert its response.
+* `HttpRequest`: Execute an HTTP request against an API endpoint and assert the response.
   Used typically to test API Gateway resources.
 
 By default, these assertions will check that the underlying call succeeds. That behavior can be extended
 to also include verification of the response.
 
 ```ts
-test.assertAws(
-  s3.getObject({ ... }).returns({ Data: ... });
-);
-
-test.invokeLambda(...).returns({ ... });
-test.invokeLambda(...).throws({ ... });
+new AwsAssertion(test, 'GetObject', {
+  request: AwsAssertionCall.S3.getObject({ ... }),
+  returns: { Data: ... },
```

Comment:
to support the use case of allowing chained calls (call 1: `StartExecution`, call 2: `GetExecutionStatus(executionId = ...)`), can we use the `Capture` model?

Reply:
Let's not focus too deeply on this. We can get into a discussion about the exact API modelling when we get to implement (and implementation design of) this.
We can get as innovative on this as needed over time or even build service specific Assertion classes that do chaining, etc.

The idea here is to capture the overall direction in terms of API, to see there are no major gaps, or innovation blockers.

I would like to focus this RFC on areas around the designs of the test execution framework, assertion model, interaction between CLI and the test frameworks, etc.

Reply:
Sure – this was I guess a poorly phrased way of saying: I would like to see first class support for call chaining in this design. I don't think it's explicitly mentioned but I hope it will be, since I think it's integral to the use of the framework.

Reply:
Ah, now I see what you mean.

This feels like the feature of the aws-assertions "module". If we look at the aws assertions as its own piece separate from the core testing system, we need to build some form of chaining (something like `Capture` perhaps as you previously alluded to). It's not necessarily something any other assertion "module" might require.

So it feels to me that this is part of the 'aws assertions' part of the system, and not something first class in the 'assertion' part of the system. But I'm also having a hard time judging that at this level - need to get my hands dirty to see how this will pan out.

Does that make sense?
Or do you see a piece that we need to add to the 'assertion' part of the system that will enable this for aws assertions?

Reply:
I guess I don't yet see the dividing line between "core" and "aws" assertions – I'm sure this is in the works as the RFC takes shape! However, it seems to me that regardless of the [infrastructure manager (AWS/Azure/...)/call provider (AWS SDK/HTTP/...)], call chaining would be very useful. Not being able to track "state" in the assertions seems to really restrict what kind of tests the library can support. But, again, this is based just off my assumptions about core vs. aws. Excited to see this progress!

Reply:
I've addressed this in the latest change.